python/pyspark/sql/pandas/conversion.py

```diff
@@ -394,10 +394,11 @@ def _create_from_pandas_with_arrow(self, pdf, schema, timezone):
 
         # Create the Spark schema from list of names passed in with Arrow types
         if isinstance(schema, (list, tuple)):
-            arrow_schema = pa.Schema.from_pandas(pdf, preserve_index=False)
+            inferred_types = [pa.infer_type(s, mask=s.isna(), from_pandas=True)
```

Comment:
So without this change, `pa.Schema.from_pandas` cannot handle pandas extension types and pd.NA values?

Reply:
pyarrow < 0.17.0 cannot handle either ([ARROW-8159](https://issues.apache.org/jira/browse/ARROW-8159)). pyarrow 0.17.x works as long as the columns that contain `pd.NA` values are not `object`-dtyped, which is the case by default as of pandas 1.0.4 (cf pandas-dev/pandas#32931). `pa.infer_type` can take a mask and thus avoids trying to infer the type of `pd.NA` values, which is what causes `pa.Schema.from_pandas` to fail here.

`pa.Schema.from_pandas` returns different types from `pa.infer_type` in two cases:
1. `Categorical` arrays
    * `pa.Schema.from_pandas` returns a `DictionaryType`
    * `pa.infer_type` returns the `value_type` of the `DictionaryType`, which is what is already used to determine the Spark type of the resulting column
2. `__arrow_array__`-implementing arrays which return a specialised Arrow type (`IntervalArray`, `PeriodArray`)
    * `pa.Schema.from_pandas` returns the type of the array returned by `__arrow_array__`
    * `pa.infer_type` does not check for `__arrow_array__` and thus fails with these arrays, however these types cannot currently be converted to Spark types anyway

Neither of these cases cause regressions, which is why I propose replacing `pa.Schema.from_pandas` with `pa.infer_type` here.

Reply:
`pa.Schema.from_pandas` will return a type that is a subclass of `pa.ExtensionType`. From that instance, there is a `storage_type` that is defined, which could then be checked as a Spark supported type. Assuming the Pandas extension array implemented `__arrow_array__`, which is recommended, see https://arrow.apache.org/docs/python/extending_types.html#controlling-conversion-to-pyarrow-array-with-the-arrow-array-protocol.

Reply:
> For the second case above, so `pa.Schema.from_pandas` returns correct types in the case? When `pa.infer_type` infers the specified array types, will it just throw error or return a wrong array type?

`pa.infer_type` will throw an error for these arrays.